Refactor: Convert Title widget to StatefulWidget

kazbeksultanov opened the pull request:
Fix: #176008

The Title widget was previously a `StatelessWidget`. This change converts it to a `StatefulWidget` to allow it to update the application switcher description when its `title` or `color` properties change. 

- The `_updateChrome` method is introduced to encapsulate the call to `SystemChrome.setApplicationSwitcherDescription`.
- `initState` now calls `_updateChrome` to set the initial description.
- `didUpdateWidget` checks if `title` or `color` have changed and calls `_updateChrome` accordingly.
- The `build` method now simply returns `widget.child`.
- Property access in `debugFillProperties` is updated to use `widget.propertyName`.

<!--
Thanks for filing a pull request!
Reviewers are typically assigned within a week of filing a request.
To learn more about code review, see our documentation on Tree Hygiene: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
-->

*Replace this paragraph with a description of what this PR is changing or adding, and why. Consider including before/after screenshots.*

*List which issues are fixed by this PR. You must list at least one issue. An issue is not required if the PR fixes something trivial like a typo.*

*If you had to change anything in the [flutter/tests] repo, include a link to the migration guide as per the [breaking change policy].*

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I signed the [CLA].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I added new tests to check the change I am making, or this PR is [test-exempt].
- [x] I followed the [breaking change policy] and added [Data Driven Fixes] where supported.
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

**Note**: The Flutter team is currently trialing the use of [Gemini Code Assist for GitHub](https://developers.google.com/gemini-code-assist/docs/review-github-code). Comments from the `gemini-code-assist` bot should not be taken as authoritative feedback from the Flutter team. If you find its comments useful you can update your code accordingly, but if you are unsure or disagree with the feedback, please feel free to wait for a Flutter team member's review for guidance on which automated comments should be addressed.

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#overview
[Tree Hygiene]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md
[test-exempt]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#tests
[Flutter Style Guide]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md
[Features we expect every widget to implement]: https://github.com/flutter/flutter/blob/main/docs/contributing/Style-guide-for-Flutter-repo.md#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/blob/main/docs/contributing/Tree-hygiene.md#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/blob/main/docs/contributing/Chat.md
[Data Driven Fixes]: https://github.com/flutter/flutter/blob/main/docs/contributing/Data-driven-Fixes.md


[![talabat.com](https://img.shields.io/badge/talabat.com-contributions-FF5A00?style=flat&logo=flutter&logoColor=white)](https://www.talabat.com) [![Talabat Flutter PRs](https://img.shields.io/badge/Talabat_Flutter_PRs-10%20merged-97ca00?style=flat&logo=flutter&logoColor=white)](https://github.com/search?q=org%3Aflutter+talabat&type=pullrequests)

kazbek-sultanov-t committed:
Refactor: Convert Title widget to StatefulWidget

The Title widget was previously a StatelessWidget. This change converts it to a StatefulWidget to allow it to update the application switcher description when its `title` or `color` properties change.

- The `_updateChrome` method is introduced to encapsulate the call to `SystemChrome.setApplicationSwitcherDescription`.
- `initState` now calls `_updateChrome` to set the initial description.
- `didUpdateWidget` checks if `title` or `color` have changed and calls `_updateChrome` accordingly.
- The `build` method now simply returns `widget.child`.
- Property access in `debugFillProperties` is updated to use `widget.propertyName`.

kazbek-sultanov-t committed:
Improve docs for Title widget

kazbek-sultanov-t committed:
Move super calls to top of method in Title widget

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

justinmc reviewed (commented):
I'm on board with making this change, but please add a test that verifies that     SystemChrome.setApplicationSwitcherDescription is called only when title or color change.

kazbek-sultanov-t commented on packages/flutter/lib/src/widgets/title.dart:
```diff
@@ -9,7 +9,11 @@ import 'basic.dart';
 
 /// A widget that describes this app in the operating system.
-class Title extends StatelessWidget {
+///
+/// `SystemChrome.setApplicationSwitcherDescription` is called once on
+/// `initState` in `StatefulWidget` and each time only when [title] and/or
+/// [color] is changed
```
Actually, you are right. We can keep as it was previously.

kazbeksultanov committed:
Remove redundant comment from Title widget

The documentation for the `Title` widget was describing its implementation details. This change removes the redundant comment.

Additionally, new tests are added to verify that `SystemChrome.setApplicationSwitcherDescription` is called only when the `title` or `color` properties actually change, and not on every rebuild with the same values.

kazbeksultanov commented:
@justinmc I reverted the doc comment on Title Widget and added tests.

chunhtai reviewed (approved):
LGTM, I saw the comment in original issue, this approach LGTM

kazbeksultanov commented on packages/flutter/lib/src/widgets/title.dart:
```diff
@@ -32,17 +32,39 @@ class Title extends StatelessWidget {
+  }
+
+  void _updateChrome() {
     SystemChrome.setApplicationSwitcherDescription(
```
Now `setApplicationSwitcherDescription ` will be called only when `title` or `color` will be changed, not on every rebuild.

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Fix failing test by correcting log indices

kazbeksultanov commented:
@chunhtai what should I do next?

kazbek-sultanov-t commented:
@benthillerkus @chunhtai @chinmaygarde anyone could add the `autosumbit` label, please 🙏 ?

kazbeksultanov commented:
@benthillerkus @chunhtai @chinmaygarde anyone could add the autosumbit label, please 🙏 ?

justinmc commented on packages/flutter/test/widgets/title_test.dart:
```diff
@@ -50,4 +50,75 @@ void main() {
+      ) async {
+        log.add(methodCall);
+        return null;
+      });
```
Can you also clean this up in a teardown after setting it?

```dart
      addTearDown(() {
        TestDefaultBinaryMessengerBinding.instance.defaultBinaryMessenger.setMockMethodCallHandler(
          SystemChannels.platform,
          null,
        );
      });
```

justinmc reviewed (commented):
Missing a teardown in the tests, otherwise this looks good.

Sorry for the slow response here, we usually check PRs once per week. Feel free to ping me on the Flutter Discord once this is ready for re-review if I'm not seeing it.

kazbeksultanov committed:
Add tearDown for mock method call handler in Title widget tests

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

justinmc reviewed (approved):
LGTM 👍 . Thanks for following up! I see some failures, but they look like flakes. I'm rerunning them.

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates

kazbeksultanov committed:
Merge branch 'master' into fix-title-widget-multiple-discription-updates
